histogram/src/histogram.rs

```diff
@@ -0,0 +1,430 @@
+// Copyright 2022 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use crate::*;
+
+use core::sync::atomic::AtomicU32;
+use core::sync::atomic::Ordering;
+
+/// A `Histogram` groups recorded values into buckets of similar values and
+/// tracks counts for recorded values that fall into those ranges.
+#[allow(non_snake_case)]
+#[allow(dead_code)]
+pub struct Histogram {
+    m: u32,
+    r: u32,
+    n: u32,
+
+    M: u64,
+    R: u64,
+    N: u64,
+    G: u64,
+
+    buckets: Box<[AtomicU32]>,
+}
+
+/// A `Builder` allows for constructing a `Histogram` with the desired
+/// configuration.
+pub struct Builder {
+    m: u32,
+    r: u32,
+    n: u32,
+}
+
+impl Builder {
+    pub fn build(self) -> Histogram {
+        Histogram::new(self.m, self.r, self.n)
+    }
+
+    /// Sets the width of the smallest bucket in the `Histogram`.
+    ///
+    /// As the `Histogram` uses base-2 internally, the resolution will be the
+    /// largest power of two that is less than or equal to the provided value.
+    /// For example, if the minimum resolution is set to 10, the width of the
+    /// smallest bucket will be 8.
+    pub fn min_resolution(mut self, width: u64) -> Self {
+        self.m = 64 - width.leading_zeros();
+        self
+    }
+
+    /// Sets the maximum value that the minimum resolution extends to.
+    ///
+    /// This value should be greater than the minimum resolution. If the value
+    /// provided is not a power of two, the smallest power of two that is larger
+    /// than the provided value will be used.
+    pub fn min_resolution_range(mut self, value: u64) -> Self {
+        self.r = 64 - value.next_power_of_two().leading_zeros();
+        self
+    }
+
+    /// Sets the maximum value that can be recorded into the `Histogram`.
+    ///
+    /// If the value provided is not a power of two, the smallest power of two
+    /// that is larger than the provided value will be used.
+    pub fn maximum_value(mut self, value: u64) -> Self {
+        self.n = 64 - value.next_power_of_two().leading_zeros();
+        self
+    }
+}
+
+impl Histogram {
+    #[allow(non_snake_case)]
+    pub fn new(m: u32, r: u32, n: u32) -> Self {
+        if r <= m || r > n || n > 64 {
+            panic!("invalid histogram config");
```

Comment:
This should either return an error or it should include a message explaining why the config is invalid.

If you decide to keep the panic then it should also be documented under the `# Panics` header in the doc comment.